cpp/arcticdb/version/python_bindings.cpp

```diff
@@ -27,6 +27,46 @@
 
 namespace arcticdb::version_store {
 
+[[nodiscard]] static std::pair<timestamp, timestamp> compute_first_last_dates(
+    timestamp start,
+    timestamp end,
+    timestamp rule,
+    ResampleBoundary closed_boundary_arg,
+    timestamp offset
+) {
+    timestamp start_with_offset = (start - offset) % rule;
+    timestamp end_with_offset = (end - offset) % rule;
+    if (closed_boundary_arg == ResampleBoundary::RIGHT) {
+        start_with_offset = start_with_offset > 0 ? start - start_with_offset : start - rule;
+        end_with_offset = end_with_offset > 0 ? end + (rule - end_with_offset) : end;
+    } else {
+        start_with_offset = start_with_offset > 0 ? start - start_with_offset : start;
+        end_with_offset = end_with_offset > 0 ? end + (rule - end_with_offset) : end + rule;
+    }
+    return {start_with_offset, end_with_offset};
+}
+
+std::vector<timestamp> generate_buckets(
+    timestamp start,
+    timestamp end,
+    std::string_view rule,
+    ResampleBoundary closed_boundary_arg,
+    timestamp offset
+) {
+    const timestamp rule_ns = [&rule]() {
+        py::gil_scoped_acquire acquire_gil;
+        return python_util::pd_to_offset(rule);
+    }();
```

Comment:
Adding a lambda just to add a scope seems less clean to me than
```
timestamp rule_ns;
{
    py::gil_scoped_acquire acquire_gil;
    rule_ns = python_util::pd_to_offset(rule);
}
```

Reply:
That's trick from Herb Sutter for initialization of constant complex variables from Herb Sutter: https://herbsutter.com/2013/04/05/complex-initialization-for-a-const-variable/ I don't mind changing it if you insist.

Reply:
Yeh keeping it `const` is nice, happy to keep it as a lambda